giraffe/src/utils/unzip.ts

```diff
@@ -0,0 +1,33 @@
+import {isLength} from './isLength'
+
+const isArrayLikeObject = value => {
+  return (
+    typeof value === 'object' &&
+    value !== null &&
+    typeof value !== 'function' &&
+    isLength(value.length)
+  )
+}
+
+const baseProperty = key => {
+  return object => (object == null ? undefined : object[key])
+}
+
+export const unzip = (array: Array<any>) => {
+  if (!(Array.isArray(array) && array.length)) {
+    return []
+  }
+  let length = 0
+  array = array.filter(group => {
+    if (isArrayLikeObject(group)) {
+      length = Math.max(group.length, length)
+      return true
+    }
+  })
+  let index = -1
+  const result = new Array(length)
+  while (++index < length) {
+    result[index] = array.map(baseProperty(index))
+  }
+  return result
```

Comment:
another example where a `do while` would be cleaner i think

Reply:
As written, it can short-circuit when length is 0. A do-while will not short-circuit.